### ISSUE TITLE
[HOLD for payment 2023-06-07] [$1000] Personal details - DOB field does not get updated if year only changed.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#16514]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/16474

## Action Performed:
1. Open settings > profile > personal details > date of birth > date of birth input
2. Press year to select the new year from the list
3. Tap on Save

## Expected Result:
DOB should be updated of the year was changed.

## Actual Result:
DOB field does not get updated if year only changed.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.89.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/227733045-de7fc375-4df9-4dae-820a-d6e51fee16ed.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0124a630171dea7ddd</li>
        <li>Upwork Job ID: 1642120128439169024</li>
        <li>Last Price Increase: 2023-04-11</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @lschurr (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by lschurr** [comment]
Reproduced. Adding Eng.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by Christinadobrzyn** [comment]
Hey @roryabraham can you take a peek at these two related jobs to see if they can be combined into this one? If they can be combined, I _think_ we can still pay the reporter for finding these, correct?

https://github.com/Expensify/App/issues/16529
https://github.com/Expensify/App/issues/16518

**Comment by lschurr** [comment]
@roryabraham - thoughts on this one?

**Comment by roryabraham** [comment]
@Christinadobrzyn @lschurr I think this looks related but different to those other two issues.

I see what @bernhardoj is saying but disagree on this point:

> We can't guarantee the currently selected day are available in the newly selected year/month

Because we can check this quite easily.

So I think we should update the Expected Results:

**Action Performed:** Change just the year in your birth date setting.

**Expected result:** When fields in the date picker are changed, validate that the new date is valid. If it is not, disable the `Save` button and display a red error above the button saying something to the effect of "Please enter a valid date"

**Comment by bernhardoj** [comment]
Yes, we need to check it. What I mean by "guarantee" is setting the new date as it is 😅. For the expected result, I think it's better to select the last available day if it's invalid. This is the behavior with native web date picker.

https://user-images.githubusercontent.com/50919443/229173947-15925e5d-eed5-442b-a4a6-61b9eaf46240.mov

**Comment by lschurr** [comment]
@roryabraham should we add the Internal or External label and work toward fixing this?

**Comment by roryabraham** [comment]
Okay, we've got a majority consensus in [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1680281888942749) on the expected behavior here:

> Action Performed: Change just the year in your birth date setting.
>
> Expected result: When fields in the date picker are changed, validate that the new date is valid. If it is not, disable the Save button and display a red error above the button saying something to the effect of "Please enter a valid date"

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0124a630171dea7ddd

**Comment by MelvinBot** [comment]
Triggered auto assignment to @greg-schroeder (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Current assignee @roryabraham is eligible for the External assigner, not assigning anyone new.

**Comment by alitoshmatov** [comment]
@hoangzinh I tried the same approach, I am wondering how you going to handle showing error message when invalid date is selected? When I tried to set year of moment object it automatically switched to 28th Feb from 29th. Also I couldn't figure out how to show error message since it is delivered from props

**Comment by KhalilMohammad** [comment]
**Proposal**
<hr />

**Please re-state the problem that we are trying to solve in this issue.**
Personal details - DOB field does not get updated if year only changed

**What is the root cause of that problem?**
When year is selected in the Select Year page, it navigates back to the DOB page. Although we have a new selectedYear but we haven't updated date field with new selected year to Form

**What changes do you think we should make in order to solve the problem?**
We need to add logic in CalenderPicker.
1. When year is changed ```setDate``` method is not triggered.
2. We can check ```selectedYear``` is not empty.
We can also check if date is valid after update year using ```isValid``` method of ```moment```
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L27
We can then trigger ```setDate``` from here. We need to add conditional logic for hidePicker and blur.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/NewDatePicker/index.js#L79
This logic can be added in ```ComponentDidMount``` or ```Constructor``` of ```CalenderPicker.js``` as we get Navigation Back from ```YearPickerPage```. ```ComonentDidUpdate``` is never going to triggered.
3. We can remove ```ComonentDidUpdate``` as it's not used functionally. We are not doing state updates, we are updating it through props.

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
DOB field does not update when we change year/month from the date picker.

### What is the root cause of that problem?
We are not updating the text input value when the month/year changed, only when we select the day.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L82-L85
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/NewDatePicker/index.js#L79-L85

`onInputChange` is the one that is responsible to update the text input value.
### What changes do you think we should make in order to solve the problem?

If we look at the month arrow press callback code, we can know that the local `currentDateView` state value will be changed, but not with the text input value.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L87-L93

(1) To reflect the update to the text input, we should call `onDayPressed` (I think we can rename it to `onSelected`) every time `currentDateView` state changes. For the `day` argument, we will pass `this.state.currentDateView.getDate()` to get the current selected date. We also need to pass a new argument to decide whether we want to hide the picker when call `onSelected`. We only want to hide the picker when we select the day. 

While working on this, I found that it's currently not possible to show an invalid date from what we expect here https://github.com/Expensify/App/issues/16514#issuecomment-1492936726. Moment already automatically handle to select the closest date when we subtract or add the current date. We can refactor the `moveToPrevMonth` & `moveToNextMonth` to manually do the date subtraction and addition, but I don't think it is worth the effort. There will be no space for the error message too as it will be beneath the picker.
<img width="353" alt="image" src="https://user-images.githubusercontent.com/50919443/229296593-303e5f88-ec11-4652-a86e-b97e40408c29.png">

So far, it works fine when we change the month. However, it does not work when we change the year because `CalendarPicker.componentDidMount` is never called, only on web. So, (1) logic is not applicable for web. It does not work because when we open the year picker, it will call the wrapper `onBlur` here which will hide the picker. When we get back to the previous page, `CalendarPicker` will be re-mounted and call `componentDidMount` instead.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/NewDatePicker/index.js#L118-L125

To handle the year changes, we need to handle it inside `NewDatePicker`. First, add a `componentDidUpdate` and if the `props.selectedYear` change and is not an empty string, update the whole date by directly call `setDate` with the new date constructed from the current date altered by the new `props.selectedYear`.

(2)
```
const newDate = moment(this.state.selectedDate);
newDate.set('year', this.props.selectedYear);
this.setDate(this.getNewDate(newDate.toDate()), false); // false to not hide the picker
```

If you look at the code above, there is a new function used which is `this.getNewDate` and this is the last thing we need to handle. Our date picker has a min and max date. To make sure the new date is between those constraint, `this.getNewDate` is made. The logic is taken from `CalendarPicker`.
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L32-L36

`this.getNewDate` will be used in both `NewDatePicker` (above [2]) and `CalendarPicker` (below).
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L87-L93
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/components/CalendarPicker/index.js#L62
It will look like this
`this.setState(prev => ({currentDateView: this.getNewDate(moment(prev.currentDateView).subtract(1, 'M').toDate())}));`

High level summary of the changes:
1. Update text input value every time day/month/year changes.
2. Constraint the selected date between min and max date.

**Result**

https://user-images.githubusercontent.com/50919443/229297601-25689cb6-ca02-414d-bf97-5b57c2db51f8.mov

**Comment by Santhosh-Sellavel** [comment]
@Christinadobrzyn It's not fixed here. Also, I left a comment [there](https://github.com/Expensify/App/issues/16529#issuecomment-1494917336). 
That doesn't look like a bug @roryabraham can you confirm that thanks!

**Comment by Santhosh-Sellavel** [comment]
Will revisit this again tomorrow!

**Comment by Santhosh-Sellavel** [comment]
There could be other options, will wait for more proposals.

**Comment by s77rt** [comment]
Looks more like a feature request than a bug. I don't have a proposal here just slightly simplifying @bernhardoj's proposal since we won't have to deal with calendar hide/show functionality, it will always be visible once https://github.com/Expensify/App/pull/16941 is merged.

<details>
<summary>1. Define clampDate to avoid repeated logic</summary>

```js
    /**
     * @param {Date} date
     * @returns {Date}
     */
    clampDate(date) {
        if (this.props.maxDate < date) {
            return this.props.maxDate;
        }
        if (this.props.minDate > date) {
            return this.props.minDate;
        }
        return date;
    }
```
</details>

<details>
<summary>2. Call onSelected whenever currentDateView is updated</summary>

```js
    componentDidUpdate(prevProps) {
        // Check if selectedYear has changed
        if (this.props.selectedYear === prevProps.selectedYear) {
            return;
        }

        // If the selectedYear prop has changed, update the currentDateView state with the new year value
        this.setState(prev => ({currentDateView: this.clampDate(moment(prev.currentDateView).set('year', this.props.selectedYear).toDate())}), () => {
            this.props.onSelected(this.state.currentDateView);
        });
    }
```

```js
    /**
     * Calls the onSelected function with the selected date.
     * @param {Number} day - The day of the month that was selected.
     */
    onDayPressed(day) {
        this.setState(prev => ({currentDateView: this.clampDate(moment(prev.currentDateView).date(day).toDate())}), () => {
            this.props.onSelected(this.state.currentDateView);
        });
    }

    moveToPrevMonth() {
        this.setState(prev => ({currentDateView: this.clampDate(moment(prev.currentDateView).subtract(1, 'M').toDate())}), () => {
            this.props.onSelected(this.state.currentDateView);
        });
    }

    moveToNextMonth() {
        this.setState(prev => ({currentDateView: this.clampDate(moment(prev.currentDateView).add(1, 'M').toDate())}), () => {
            this.props.onSelected(this.state.currentDateView);
        });
    }
```
</details>

**Comment by lschurr** [comment]
@Santhosh-Sellavel - are we still waiting for more proposals on this one?

**Comment by MelvinBot** [comment]
@lschurr @roryabraham @Santhosh-Sellavel this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by lschurr** [comment]
@Santhosh-Sellavel @roryabraham - thoughts on this one?

**Comment by roryabraham** [comment]
I am kind of thinking that @ArekChr should work on this, because they implemented the component as part of a feature request, and it's still pretty new. I'm not sure we should have a separate bug bounty for it.

**Comment by ArekChr** [comment]
Hey, @roryabraham I can work on that feature

**Comment by MelvinBot** [comment]
📣 @ArekChr You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
⚠️ This issue has had its price increased by 4x or more. Please review the issue and ensure the price is correct.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by burczu** [comment]
Hi I'm Bartek from Callstack - expert contributor group - since @ArekChr is on a longer sick leave now, I'll take over this issue from him

**Comment by MelvinBot** [comment]
📣 @burczu You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
@burczu @lschurr @roryabraham @Santhosh-Sellavel this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by MelvinBot** [comment]
@burczu, @lschurr, @roryabraham, @Santhosh-Sellavel Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by burczu** [comment]
Daily update: I had limited time to work on this for now - I've got familiar with the issue and will start working on the proposal.

**Comment by burczu** [comment]
Update: I've finally finished working on another issue and started working on proposal for this one.

**Comment by MelvinBot** [comment]
@burczu @lschurr @roryabraham @Santhosh-Sellavel this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @Santhosh-Sellavel is eligible for the Internal assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
@burczu, @lschurr, @roryabraham, @Santhosh-Sellavel Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by burczu** [comment]
I've read the whole thread, all the proposals and the slack conversation and the more I think about it the more I conclude that this issue is just invalid...

Look at how this date picker works right now: we have an input field above the calendar that presents the selected date. The only way to provide the date right now is to select the day from the calendar, and **the row with year and month selectors in the calendar is just navigation** for the calendar component. They just change what is shown in the calendar, not the selected date and I think it is correct behaviour: 

https://user-images.githubusercontent.com/7771498/233978113-be598c02-1d7b-4181-b4a7-3c2ad3e56799.mp4

I've checked how it's done in google calendar as an example:

https://user-images.githubusercontent.com/7771498/233976907-f05ed29a-e5bd-41ab-9eeb-4fb4e6b21e1c.mp4

It actually works similar to how it works in Expensify (with the exception that it allows only to change the month), but changing the month does not change the selected date. So my conclusion is that changing the year nor the month should not change the selected date whatsoever.

What do you think?

**Comment by burczu** [comment]
@Santhosh-Sellavel Yes, I've already read all of this - but I think I have different concern than @bernhardoj in his comment. In my opinion we are changing the common behaviour of the date picker. But of course I will provide the proposal for the expected changes described in this issue. Please tell me how the date picker should behave in the situation I've described - I think this is case is not covered by the expected behaviour description:

> For example, let's say the user selects the date: 1984-02-29. Then he changes the year to 1983 and, if we allow this, he 
sees incorrect date: 1983-02-29 in the input field... Now, what should he see in the calendar component? No day selected?

Unselecting the date in the calendar seem to be the best option here, right?

**Comment by roryabraham** [comment]
> For example, let's say the user selects the date: 1984-02-29. Then he changes the year to 1983 and, if we allow this, he
sees incorrect date: 1983-02-29 in the input field... Now, what should he see in the calendar component? No day selected?

1. Day is unselected in the calendar
2. An error message displays saying "Invalid Date"

**Comment by Santhosh-Sellavel** [comment]
@burczu Are we good here?

**Comment by burczu** [comment]
@Santhosh-Sellavel Yes - I've got the whole solution and now preparing the proposal. Should be posted here tomorrow.

**Comment by burczu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The text input above the calendar picker in the date picker component is not updated if only a year or month has changed in the calendar picker.

### What is the root cause of that problem?

Right now, it is not possible to set an incorrect date in the date picker. This is just how it's designed and the issue is actually a feature request.

To change this feature and start updating the text input not only when user picks the day from the calendar but also when he has changed the year or month, we need to allow providing incorrect date in the date picker. It's because changing a year or month may end up setting a date that is out of range, e.g.: user selects 1984-02-29 (1984 was a leap year), then he changes the year to 1983 and ends up with the date of 1983-02-29 which is incorrect because February had only 28 days in 1983.

### What changes do you think we should make in order to solve the problem?

I think we can divide the solution into three steps:

#### Switch to use strings

To allow providing incorrect date in the date picker we need to switch from operating on dates to oparating on strings, because it's impossible to initialize `moment` (or `Date`) object with an incorrect date (it automatically changes it into the closest correct one).

Let's start from PropTypes, we need to change `datePickerPropTypes` file:

https://github.com/Expensify/App/blob/1c749c07cf307cf9689fbc1d9794f74a58ab935e/src/components/NewDatePicker/datePickerPropTypes.js#L9-L32

like this:

```javascript
const propTypes = {
    ...fieldPropTypes,

    value: PropTypes.string,
    
    defaultValue: PropTypes.string,

    ... // other values
};
```

And reflect these changes in the `DatePicker` component itself:

https://github.com/Expensify/App/blob/1c749c07cf307cf9689fbc1d9794f74a58ab935e/src/components/NewDatePicker/index.js#L25-L28

https://github.com/Expensify/App/blob/1c749c07cf307cf9689fbc1d9794f74a58ab935e/src/components/NewDatePicker/index.js#L50-L58

should be changed like below:

```javascript
constructor(props) {
   ...
   
   this.state = {
            selectedMonth: null,
            selectedDate: props.value || props.defaultValue || undefined,
        };
}

...

setDate(selectedDate) {
    this.setState({selectedDate}, () => {
        this.props.onInputChange(selectedDate)
    })
}
```

We need to do the same with the `newCalendarPickerPropTypes` file:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/calendarPickerPropTypes.js#L5-L29

should be changed like below:

```javascript
const propTypes = {
    /** An initial value of date */
    value: PropTypes.string,
    
    ... // the rest of props stays the same
}
```

And the change in the `CalenderPicker` components constructor:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L19-L43

Now that the `props.value` is a `string` not a `Date` we need to transform it into date manually when initializing the `currentDateView` variable:

```javascript
constructor(props) {
    super(props);
    
    let currentDateView = moment(props.value).toDate();
    
    ... // the rest stays the same
}
```

#### Change the calendar picker logic

Right now, in the `CalendarPicker` we have a `currentDateView` date object stored in the state that is responsible for keeping two things at the same time: the selected date and what should be shown in the calendar. I think we should distinguish these two responsibilities - this way we will be able to show the incorrect date in the text input, and have calender in an "unselected" state at the same time.

Let's introduce three new state values in the constructor:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L35-L37

will be changed like below:

```javascript
const currentSelection = moment(props.value).format(CONST.DATE.MOMENT_FORMAT_STRING);
const selectionParts = currentSelection.split('-');

this.state = {
    currentDateView,
    selectedYear: selectionParts[0],
    selectedMonth: selectionParts[1],
    selectedDay: selectionParts[2],
};
```

Now we can move on to change the logic behind selecting the day, month and year. First, let's introduce `clampDate` method to avoid repeating (it's taken from two other proposals):

```javascript
/**
 * @param {Date} date
 * @returns {Date}
 */
clampDate(date) {
    if (this.props.maxDate < date) {
        return this.props.maxDate;
    }
    if (this.props.minDate > date) {
        return this.props.minDate;
    }
    return date;
}
```

I would also introduce another helper method that returns a date string in the format of `YYYY-MM-DD` combined using current state values - we will need it in our handlers.

```javascript
getSelectedDateString() {
    return `${this.state.selectedYear}-${this.state.selectedMonth}-${this.state.selectedDay}`;
}
```

Then we can change the `onDayPressed` handler from:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L75-L82

to this:

```javascript
/**
 * Calls the onSelected function with the selected date.
 * @param {Number} day - The day of the month that was selected.
 */
onDayPressed(day) {
    this.setState((prev) => {
        const momentDate = moment(prev.currentDateView).date(day);
        const clampedDate = this.clampDate(momentDate.toDate());
        return {
            selectedDay: `${day < 10 ? `0${day}` : day}`,
            currentDateView: clampedDate,
        };
    }, () => {
        this.props.onSelected(this.getSelectedDateString());
    });
}
```

Setting the `currentDateView` state value is actually the same as proposed by @s77rt in his [comment](https://github.com/Expensify/App/issues/16514#issuecomment-1499863159). What is added here is to set `selectedDay` value separately. 

Please also note how the callback has changed - `this.props.onSelected(this.getSelectedDateString())` will pass the date string created from newly added `selectedSomethig` state values and not from `currentDateView`.

For month handlers we can add very similar changes:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L84-L90

will be changed like below:

```javascript
moveToPrevMonth() {
    this.setState((prev) => {
        const momentDate = moment(prev.currentDateView).subtract(1, 'M');
        const clampedDate = this.clampDate(momentDate.toDate());

        const month = clampedDate.getMonth() + 1;
        const year = clampedDate.getFullYear(); // year might have changed too

        return {
            selectedMonth: `${month < 10 ? `0${month}` : month}`,
            selectedYear: `${year}`,
            currentDateView: clampedDate,
        };
    }, () => {
        this.props.onSelected(this.getSelectedDateString());
    });
}

moveToNextMonth() {
    this.setState((prev) => {
        const momentDate = moment(prev.currentDateView).add(1, 'M');
        const clampedDate = this.clampDate(momentDate.toDate());

        const month = clampedDate.getMonth() + 1;
        const year = clampedDate.getFullYear(); // year might have changed too

        return {
            selectedMonth: `${month < 10 ? `0${month}` : month}`,
            selectedYear: `${year}`,
            currentDateView: clampedDate,
        };
    }, () => {
        this.props.onSelected(this.getSelectedDateString());
    });
}
```

Now we can finally move on to handle year changes. First, small adjustment in the `onYearPickerPressed` handler:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L67-L73

It's good to change the source of the `currentYear` here, to keep everything consistent:

```javascript
onYearPickerPressed() {
    ...
    const currentYear = parseInt(this.state.selectedYear, 10);
    ...
}
```

Then, we need to change the `componentDidUpdate` method implementation from:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/components/CalendarPicker/index.js#L52-L60

into this (very similarily to changes made to day and month handlers):

```javascript
componentDidUpdate(prevProps) {
    // Check if selectedYear has changed
    if (this.props.selectedYear === prevProps.selectedYear) {
        return;
    }

    // If the selectedYear prop has changed, update the currentDateView state with the new year value
    this.setState((prev) => {
        const newMomentDate = moment(prev.currentDateView).set('year', this.props.selectedYear);
        const clampedDate = this.clampDate(newMomentDate.toDate());

        return {
            selectedYear: this.props.selectedYear,
            currentDateView: clampedDate,
        };
    }, () => {
        this.props.onSelected(this.getSelectedDateString());
    });
}
```

Thanks to separating `currentDateView` from `selectedYear`, `selectedMonth` and `selectedDay` when we have e.g. date 1984-02-29, and change the year to 1983, the `selectedMonth` stays set to `02`, the `selectedDay` stays set to `29` so the date in the text input will be `1983-02-29`. At the same time, the `currentDateView` may be automatically changed to `1983-02-28` because it is created using moment.js, but it used only to show correct year and month in the calendar picker.

#### Fix validation

The last thing is to adjust validation. Actually it is already set up, but the validation message is hidden because the `CalendarPicker` component is positioned absolutely. So the only change we need here is to remove it and add some margin:

https://github.com/Expensify/App/blob/243e791d7ebe26f8e251b9b4d7727e940bea14dd/src/styles/styles.js#L3103-L3110

should be changed like below:

```css
datePickerPopover: {
    backgroundColor: themeColors.appBG,
    width: '100%',
    alignSelf: 'center',
    zIndex: 100,
    marginTop: 8,
},
```

### What alternative solutions did you explore? (Optional)

As I wrote in my [comment](https://github.com/Expensify/App/issues/16514#issuecomment-1519916340) the alternative is to re-think this feature request because, in my opinion, the current behaviour is just correct and the calendar picker implementation works like most similar calendar pickers we can find in the internet.

--

Screen recording of how it works after the above changes:

https://user-images.githubusercontent.com/7771498/235112516-72ccf6f2-068d-4aab-b5f6-965f5fe2a4c1.mp4

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by lschurr** [comment]
@Santhosh-Sellavel could you review this proposal?

**Comment by Santhosh-Sellavel** [comment]
@burczu 
Reminder: Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by lschurr** [comment]
Bump @roryabraham

**Comment by burczu** [comment]
@Santhosh-Sellavel thanks for your thoughts. In terms of the code quality of course it will be be done better in the PR - this is just the idea of we want to do it. And sorry for too long proposal - I'll do my best to me more brief next time!

**Comment by mountiny** [comment]
> So many code snippets, and too many details. also, the code is not DRY and could be improved. I feel this could have been written more briefly. But otherwise okay for me to will test on PR.

I think this is a bit of a double edged sword @Santhosh-Sellavel 😅  too many snippets / the code you should not write here is not dry 😂 

@Santhosh-Sellavel feel encouraged not to wait for Rory here, we got to a busy period now so it might take a while to respond. I think we can resolve the left over concerns in a PR!

@burczu Feel free to start a PR and work with @Santhosh-Sellavel to clean it up and test it well. Thank you both, great job!

**Comment by lschurr** [comment]
Thanks @mountiny!

**Comment by lschurr** [comment]
How are we doing on this one @Santhosh-Sellavel @burczu?

**Comment by burczu** [comment]
> How are we doing on this one @Santhosh-Sellavel @burczu?

Just added draft PR - we will discuss it internally at Callstack and it should be ready for review tomorrow morning.

**Comment by burczu** [comment]
@Santhosh-Sellavel the PR is ready for review

**Comment by lschurr** [comment]
@Santhosh-Sellavel - have you had a chance to review this PR yet?

**Comment by Santhosh-Sellavel** [comment]
Yes already on it.

**Comment by lschurr** [comment]
Any updates here? Looks like the PR still needs to be merged right @Santhosh-Sellavel @roryabraham?

**Comment by lschurr** [comment]
I'm heading out of office from May 17-23rd. Re-added the Bug label and kept myself assigned. If this is still open and needs action when I return on May 24th, I will take this back from you @greg-schroeder.

Latest update: waiting for PR to be merged - https://github.com/Expensify/App/pull/18577

**Comment by greg-schroeder** [comment]
Seems PR is still awaiting review - maybe you can take a look @roryabraham?

**Comment by greg-schroeder** [comment]
Bump @roryabraham

**Comment by lschurr** [comment]
Taking this back from you @greg-schroeder - thanks!

How we doing on this one @roryabraham?

**Comment by Santhosh-Sellavel** [comment]
Seems @roryabraham is OOO. 

@lschurr Can we wait or assign another CME here?

@mountiny  If you have time for this, Assuming as you already know the context here

**Comment by mountiny** [comment]
Seeing the ping just now, I can help!

**Comment by lschurr** [comment]
Had to recreate the job since it closed. @Santhosh-Sellavel and @burczu, could you apply to the job in Upwork? https://www.upwork.com/jobs/~01be498948453fb051

**Comment by lschurr** [comment]
Also @Santhosh-Sellavel - do we need a regression test for this one?

**Comment by Santhosh-Sellavel** [comment]
@lschurr Only C+ is due here. I applied for the job. @burczu is from Callstack Team.

**Comment by lschurr** [comment]
Ah, great thanks!

**Comment by lschurr** [comment]
Paid and closed.
